docs/api/utils/proxyWithComputed.mdx

```diff
@@ -48,16 +36,60 @@ const state2 = proxyWithComputed(
 )
 
 // if you want a computed value to derive from another computed, you must declare the dependency first:
-const state = proxyWithComputed(
-  {
-    count: 1,
-  },
+const state3 = proxyWithComputed(
+  { count: 1 },
   {
     doubled: memoize((snap) => snap.count * 2),
     quadrupled: memoize((snap) => snap.doubled * 2),
   }
 )
 ```
 
-The last use case fails to infer types in TypeScript
-[#192](https://github.com/pmndrs/valtio/issues/192).
+The last use case fails to infer types in TypeScript, see [#192](https://github.com/pmndrs/valtio/issues/192).
+
+#### Disclaimer on `proxy-memoize`
+
+Be careful not to overuse `proxy-memoize` because `proxy-memoize` and `useSnapshot` do similar optimization and double optimization may lead to less performance.
+
+The similarity is that both `proxy-memoize` and `useSnapshot` keep caches of data; the difference is that getters invoked against `snapshot`s or `useSnapshot`s are not cached and will re-invoke the getter/re-calculate the result each time they are accessed; whereas `proxy-memoize` will use the `snap` argument to calculate the getter result only once per snapshot. )
+
+#### Compared to Classes
+
+Using `proxyWithComputed` is similar in spirit to using classes, e.g. the above `proxyWithComputed` could be written as:
```

Comment:
I'm not sure if I agree. Feels misleading. I would remove this section.

Reply:
This seems like the biggest thing I'm still struggling with: what are your recommendations of "when to use a class?" vs. "when to use proxyWithComputed?"

I can't tell if the difference is just personal style (some people just don't like classes, which is fine), or if there is something else...

At first, I assumed the big plus of `proxyWithComputed` was that you could use it with `proxy-memoize` to get getter caching, so that's what I wrote up as my best guess.

But since getter caching is not, per your other comment, the primary use case for `proxyWithComputed`, I'm failing to guess/observe what the trade-off would be.

So, dunno, I think that's the shortest way I can ask it: when do you recommend classes, vs. when do you recommend proxyWithComputed?

Then I'll work on rewriting this to reflect that answer. Thank you!

Reply:
Good point. OK, the benefit or `proxy-memoize` is understood. Let's see how proxyWithComputed works without `proxy-memoize`.

(btw, I wanted to drop proxyWithComputed when I added derive, but with this unique capability, I left proxyWithComputed. In general, I don't recommend proxyWithComputed, it's very hard to understand.)

Object getters and class getters are just functions. So, let's compare pure functions and proxyWithComputed.

https://codesandbox.io/s/unruffled-cdn-o82gup?file=/src/App.tsx

The thing is, as you see, snapshot works differently.
Let me know if something is unclear.

Reply:
Nice catch! Yeah, I missed that one. Let's drop proxyWithComputed, and just add a guide in docs how to use object getters (and setters), with optionally `proxy-memoize`.

> // do we need a snapshot(this) call for this to work? not sure...

Yes, we definitely do.

Reply:
#650